### DISCUSSION TITLE
I can't see HSL functions after install

### DISCUSSION THREAD

**Discussion by AndrewLister-STFC** [JuliaSmoothOptimizers/HSL.jl#170] [Q&A] [answered]
Hi, I was following the install instructions and the package doesn't seem to have any functions defined. I'm probably just missing something as I'm still getting started in Julia.

What I did:
- Download the [latest coinhsl](https://licences.stfc.ac.uk/product/coin-hsl)
- Start Julia
- `julia> using Pkg`
- `julia> Pkg.add("HSL")`
- `julia> ENV["HSL_ARCHIVES_PATH"] = "/home/example/hsl_staging/"` (which contains coinhsl-2022.11.09.tar.gz)
- `julia> Pkg.build("HSL")`
- `julia> Pkg.test("HSL")`
- exit Julia and open again

Now when I try to use it I just get:
```
julia> using HSL
julia> HSL.available_hsl_algorithms
Dict{String, VersionNumber} with 1 entry:
  "coinhsl" => v"2022.11.9"

julia> HSL.libcoinhsl
"<home_dir>/.julia/packages/HSL/QUaxT/deps/usr/lib/libcoinhsl.so"

julia> Ma97
ERROR: UndefVarError: Ma97 not defined
```

Does anyone know what I'm doing wrong?

**Comment by amontoison** [comment, marked as answer]
Hi Andrew!
The functions of the HSL packages are only defined if the related packages are installed.
If you want to use Ma97 or the other solvers, you need the archives of the following files:
https://github.com/JuliaSmoothOptimizers/HSL.jl/blob/main/deps/versions.jl

We would like to use the COIN-HSL library in the future but we need to have `c` and `z` versions of the solvers.

**Reply by AndrewLister-STFC** [reply]
Thanks Alexis, so currently the coinhsl lib is not usable for HSL.jl (unless we add the complex codes)? I'll try again with separate packages for now! Unfortunately I don't have time at the moment to work on a new collection or update the coinhsl build for complex but at least I understand the issues a bit better now.

**Reply by amontoison** [reply]
I only added the COIN-HSL library for Ipopt.jl. It's for that reason that the support of COIN-HSL is not documented yet.

**Comment by amontoison** [comment]
With the help of `Clang.jl`, we can generate wrappers for the C routines directly from the headers provided by the HSL solvers.
The script is in the folder `gen`.
The result is quite neat: https://github.com/JuliaSmoothOptimizers/HSL.jl/commit/be30241c6642c0571d3ff97cd7c467240d61668c.
The issue is that we can't use the COIN-HSL library if the symbols for complex routines are missing.

**Comment by VicenteCrutherswB** [comment]
It looks like you have successfully installed the HSL package and its dependencies, but you may need to explicitly load the HSL subpackage that contains the function you are trying to use.

In this case, Ma97 is a function in the HSL_MA97 subpackage, so you can try loading that subpackage explicitly like this:

using HSL.HSL_MA97

Then you should be able to use the Ma97 function as expected.

**Comment by amontoison** [comment]
For information, we don't have submodules for the HSL packages.
The issue here is that the file `deps/deps.jl` generated after a `pkg> build HSL` had a `libcoinshl` and we don't have any wrapper with this library.
